File: content/examples/other/leetcode/problems/detect-cycle-in-a-linked-list/example-1/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def has_cycle_hashset(head):
    seen = set()
    cur = head
    while cur:
        if cur in seen:
            return True
        seen.add(cur)
        cur = cur.next
    return False


if __name__ == "__main__":
    a = ListNode(1)
    b = ListNode(2)
    c = ListNode(3)
    a.next = b
    b.next = c
    c.next = b
    print(has_cycle_hashset(a))

File: content/examples/other/leetcode/problems/detect-cycle-in-a-linked-list/example-1/solution.js
class ListNode {
  constructor(val, next = null) {
    this.val = val;
    this.next = next;
  }
}

function hasCycleHashSet(head) {
  const seen = new Set();
  let cur = head;
  while (cur) {
    if (seen.has(cur)) return true;
    seen.add(cur);
    cur = cur.next;
  }
  return false;
}

if (require.main === module) {
  const a = new ListNode(1);
  const b = new ListNode(2);
  const c = new ListNode(3);
  a.next = b;
  b.next = c;
  c.next = b;
  console.log(hasCycleHashSet(a));
}

module.exports = { ListNode, hasCycleHashSet };
